packages/xdl/src/TurtleApi.ts

```diff
@@ -0,0 +1,200 @@
+import QueryString from 'querystring';
+
+import FormData from './tools/FormData';
+import { JSONObject, JSONValue } from '@expo/json-file';
+import axios, { AxiosRequestConfig } from 'axios';
+import concat from 'concat-stream';
+import ExtendableError from 'es6-error';
+import fs from 'fs-extra';
+import idx from 'idx';
+import merge from 'lodash/merge';
+
+import Config from './Config';
+
+
+const apiBaseUrl = `${Config.turtleApi.scheme}://${Config.turtleApi.host}:${Config.turtleApi.port}`;
+
+export class TurtleApiError extends ExtendableError {
+  code: string;
+  details?: JSONValue;
+  serverStack?: string;
+  readonly _isApiError = true;
+
+  constructor(message: string, code: string = 'UNKNOWN') {
+    super(message);
+    this.code = code;
+  }
+}
+
+type RequestOptions = {
+  httpMethod: 'get' | 'post' | 'put' | 'delete';
+  queryParameters?: QueryParameters;
+  body?: JSONObject;
+};
+
+type QueryParameters = { [key: string]: string | number | boolean | null | undefined };
+
+type TurtleApiClientOptions = {
+  sessionSecret?: string;
+};
+
+export default class TurtleApiClient {
+  sessionSecret: string | null = null;
+
+  static clientForUser(user?: TurtleApiClientOptions | null): TurtleApiClient {
+    if (user && user.sessionSecret) {
+      return new TurtleApiClient({ sessionSecret: user.sessionSecret });
+    }
+    return new TurtleApiClient(
+    );
+  }
+
+  constructor(options: TurtleApiClientOptions = {}) {
+    if (options.sessionSecret) {
+      this.sessionSecret = options.sessionSecret;
+    }
+  }
+
+  async getAsync(
+    methodName: string,
+    args: QueryParameters = {},
+    extraOptions?: Partial<RequestOptions>,
+  ) {
+    return this._requestAsync(
+      methodName,
+      {
+        httpMethod: 'get',
+        queryParameters: args,
+      },
+      extraOptions,
+    );
+  }
+
+  async postAsync(
+    methodName: string,
+    data?: any,
+    extraOptions?: Partial<RequestOptions>,
+  ) {
+    return this._requestAsync(
+      methodName,
+      {
+        httpMethod: 'post',
+        body: data,
+      },
+      extraOptions,
+    );
+  }
+
+  async putAsync(
+    methodName: string,
+    data: JSONObject,
+    extraOptions?: Partial<RequestOptions>,
+  ) {
+    return this._requestAsync(
+      methodName,
+      {
+        httpMethod: 'put',
+        body: data,
+      },
+      extraOptions,
+    );
+  }
+
+  async deleteAsync(
+    methodName: string,
+    extraOptions?: Partial<RequestOptions>,
+  ) {
+    return this._requestAsync(
+      methodName,
+      {
+        httpMethod: 'delete',
+      },
+      extraOptions,
+    );
+  }
+
+  async _requestAsync(
+    methodName: string,
+    options: RequestOptions,
+    extraRequestOptions?: Partial<RequestOptions>,
+  ) {
+    const url = `${apiBaseUrl}/${methodName}`;
```

Comment:
Please use the `url.format` function (`import url from 'url';`).
https://nodejs.org/dist/latest-v12.x/docs/api/url.html#url_url_format_url_options

Same below.